feat, refactor: add "bottom" position opt, update to Lua API command creation

mcauley-penney opened the pull request.

mcauley-penney committed:
refactor: use Lua API for user command

    • remove "plugin" dir
    • make toggle() local

mcauley-penney committed:
feat: add "bottom" as valid "position" opt

    If "bottom" is given as "position" configuration value, open
    the docs window at the bottom of the window. Adding the "height"
    option complicates setting the window dimensions, e.g. if we are
    opening a split on the sides but also change the height of the
    window, it will resize "prev_win" as well. As a consequence, we
    want to "reset" the width or height vals back to the current win
    dimension depending on the position we choose.

    Associated refactors:
        • add default "height" value in default configuration
        • rename user and default configuration tables for clarity
        • programatically set default configuration table

mcauley-penney commented:
Below is a GIF example of the "bottom" option in use. For this, I used the configuration

```lua
use({                                
    "mcauley-penney/nvim-docs-view", 
    opt = true,                      
    cmd = { "DocsViewToggle" },      
    config = function()              
        require("docs-view").setup({ 
            position = "bottom",     
        })                           
    end,                             
})                                   
```
![docs-view](https://user-images.githubusercontent.com/59481467/174680068-e7a34e2e-94d4-48a6-8ab9-2a737e07b12d.gif)

amrbashir commented on lua/docs-view.lua:
```diff
@@ -1,38 +1,35 @@
 local M = {}
-local config = {
+local default_cfg = {
```
if this is named `default_cfg` then there should be another table called just `cfg` which will be the result of the merge between `default_cfg` and `user_cfg`

mcauley-penney replied:
I agree; when I originally wrote these modifications, I attempted to use `tbl_extend`. The reason that I did not follow through with it is because 

1. I think it would require that the proposed `cfg` table is created at the module level instead of inside the setup function because the command itself relies upon a table from the module level.
2. the setup function, as it is now, already updates the `default_cfg` (or, as it is currently named, `config`) and I didn't want to change that because I figured that you wanted it to be that way

If we want to create a new table that is the result of the joining, I don't know how we'd manage to get it to the user command inside of the setup function just yet. I'll have to look into it. I don't know if user commands take default arguments that aren't from the command line.

amrbashir replied:
Here is a quick snippet of what I am thinking
```lua
local default_cfg = {
  width = 60,
  height = 10,
  position = "left", 
}
local cfg = {}

M.setup = function(user_cfg)
  cfg = vim.tbl_extend("force", default_cfg, user_cfg); -- and use `cfg` in the `toggle` function
end
```

mcauley-penney replied:
This makes sense, I don't know why I didn't recognize that I could declare at the module level and assign at the function level. Good idea 👍🏻 I will work on that now

mcauley-penney replied:
Are you comfortable with `default_cfg` being moved to the setup function? It is no longer used outside of it and that would give it the tightest scope.

amrbashir replied:
yeah that sounds good

amrbashir commented on lua/docs-view.lua:
```diff
@@ -1,38 +1,35 @@
     prev_win = vim.api.nvim_get_current_win()
 
-    if config.position == "left" then
+    if default_cfg.position == "bottom" then
+      vim.api.nvim_command("bel new")
+      default_cfg["width"] = vim.api.nvim_win_get_width(prev_win)
```
I don't like the idea of mutating the `default_cfg`, I think using a variable local to the function would be better, something like this
```suggestion
    local width = default_cfg["width"]
    if default_cfg.position == "bottom" then
      vim.api.nvim_command("bel new")
      width = vim.api.nvim_win_get_width(prev_win)
```

mcauley-penney replied:
Okay! In that case, we would want to check if `width` and `height` are defined and use them for `math.ceil` conditionally, I think, like this:

```lua
vim.api.nvim_win_set_height(win, math.ceil(height or default_cfg.height))
vim.api.nvim_win_set_width(win, math.ceil(width or default_cfg.width)) 
```

Is this what you had in mind?

amrbashir replied:
with the introduction of `cfg` from the review comment above, it will be 
```lua
local width = cfg["width"]
local height = cfg["height"]

-- and then, height/width will surely be defined because cfg is a merge of default_cfg and use_cfg
vim.api.nvim_win_set_height(win, math.ceil(height))
vim.api.nvim_win_set_width(win, math.ceil(width)) 
```

amrbashir reviewed (changes requested):
Thank you, just two small nitpicks and this should be ready for merge

amrbashir commented:
Also if you don't mind, can you update the demo gif too in the README.md?

mcauley-penney commented:
> Also if you don't mind, can you update the demo gif too in the README.md?

Yeah, no problem! When we finalize all of this I will create a GIF for the readme

mcauley-penney commented:
How do you feel about this GIF? If you'd like, I can change the font or color scheme to be something more attractive.

![doc-view-example](https://user-images.githubusercontent.com/59481467/174690571-5e964f5d-29eb-43ee-9f01-1c7d163b244e.gif)

Update: I made the same GIF over so that I could smooth out some of the movements. It has the same displays of functionality.

amrbashir commented:
it is perfect thank you

mcauley-penney committed:
refactor: cleanup after change request

    • use tbl_extend() to produce a new configuration table from
      the default cfg and user cfg

    • choose local variables over updating the default cfg when
      updating height and width inside of toggle() based on the
      chosen window position

    • move the default cfg table inside of setup()

mcauley-penney committed:
Update README.md

Add GIF displaying functionality

amrbashir committed:
include demo.gif in the repo

amrbashir reviewed (approved):
Thank you

mcauley-penney commented:
> Thank you

No problem, and thank you!
